### PR TITLE
basic composition

### LINKED ISSUE #7: Basic Composition Combinator -- Implementation
- [ ] Implement function
- [ ] Include comments
- [ ] Write tests

We have a simple implementation of this, but it only accepts exactly two Measurements, and constructs a function returning a 2-tuple. Now that we have the AnyXXX facilities, it should accept an arbitrary number of Measurements, and construct a function returning Vec<AnyObject>.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #7 
Replaces #424 

The resulting API is simplified significantly by the maps. We compose whatever privacy distances are propagated forward by the maps instead of accepting distances in the constructor API. 

Type-erased measurements emit `AnyObjects`, so the composed measurement emits a `Vec<AnyObject>`. This is nice, because each `AnyObject` can box a different type. Therefore, in Python, composition of measurements with different output domains and output carrier types is seamless. I've also adjusted the `_convert.py` code to recursively unwrap `AnyObjects`, making it possible to arbitrarily nest sequential compositors.

**An implementation quirk:**
In bindings languages, `AnyObject` is an opaque, zero-sized struct, so bindings languages can't read the `AnyObject` vector that the `FfiSlice` points to in memory. I've added an ffi helper to retrieve a vector of pointers to locations of `AnyObjects` in this vector. The helper accepts a pointer to an `FfiSlice` of `Vec<AnyObject>` and returns an `FfiSlice` of `Vec<AnyObjectPtr>`. `_convert.py` calls out to this helper whenever it unwraps a `Vec<AnyObject>`.

### CODE REVIEW

File: python/src/opendp/_convert.py [hunk 1]
```diff
@@ -204,38 +201,60 @@ def _slice_to_string(raw: FfiSlicePtr) -> str:
...
+
+    if inner_type_name == 'AnyObject':
+        from opendp._data import ffislice_of_anyobjectptrs
+        raw = ffislice_of_anyobjectptrs(raw)
+        array = ctypes.cast(raw.contents.ptr, ctypes.POINTER(AnyObjectPtr))[0:raw.contents.len]
+        res = list(map(c_to_py, array))
+        # when the top-level AnyObject is freed, it recursively frees all anyobjects inside of it
+        # adjust the type of constituent AnyObjects so that __delete__ is not called when they are dropped
+        for elem in array:
+            elem.__class__ = ctypes.POINTER(AnyObject)
```
Comment (andrewvyrros) [inline python/src/opendp/_convert.py:256]: This is a neat trick, we just need to remember the implications in potential future changes. I don't think it's possible now, but if there ever were a situation where Python could remove an AnyObject from its parent Vec, then we'd have to rethink this.

**Review by andrewvyrros — COMMENTED** [review]
For my future reference, here's my understanding of how the composed data transfer works. (Copied from a Slack conversation.)

* When Python calls `opendp_comb__make_sequential_composition_static_distances`, it sends an `AnyObject` containing a `Vec<AnyMeasurementPtr>`. `opendp_comb__make_sequential_composition_static_distances` manually unpacks this into a `Vec<&AnyMeasurement>`.
* When Rust returns from `opendp_core__measurement_invoke`, it sends an `AnyObject` containing a `Vec<AnyObject>`. Python calls `opendp_data__ffislice_of_anyobjectptrs` to turn this into an `FfiSlice` of `AnyObjectPtr`, which `_slice_to_vector` manually turns into a `List[AnyObject]`.

**Review by andrewvyrros — COMMENTED** [review]
For my future reference, here's my understanding of how the composed data transfer works. (Copied from a Slack conversation.)

* When Python calls `opendp_comb__make_sequential_composition_static_distances`, it sends an `AnyObject` containing a `Vec<AnyMeasurementPtr>`. `opendp_comb__make_sequential_composition_static_distances` manually unpacks this into a `Vec<&AnyMeasurement>`.
* When Rust returns from `opendp_core__measurement_invoke`, it sends an `AnyObject` containing a `Vec<AnyObject>`. Python calls `opendp_data__ffislice_of_anyobjectptrs` to turn this into an `FfiSlice` of `AnyObjectPtr`, which `_slice_to_vector` manually turns into a `List[AnyObject]`.

File: rust/src/ffi/any.rs [hunk 2]
```diff
@@ -130,8 +126,10 @@ pub struct AnyObject {
 
 impl AnyObject {
     pub fn new<T: 'static>(value: T) -> Self {
-        Self {
-            type_: Type::of::<T>(),
+        let type_ = Type::of::<T>();
+
+        AnyObject {
+            type_,
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/ffi/any.rs:132]: This change was a leftover from when I did some print debugging to see what the type_ was. Should have just used `dbg!`.

File: rust/src/comb/compose/mod.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+impl<Q: InfAdd + Zero + Clone> SequentialCompositionStaticDistancesMeasure
+    for ZeroConcentratedDivergence<Q>
+{
+    fn compose(&self, d_i: &Vec<Self::Distance>) -> Fallible<Self::Distance> {
+        d_i.iter().try_fold(Q::zero(), |sum, d_i| sum.inf_add(d_i))
+    }
+}
+
+pub fn make_sequential_composition_static_distances<DI, DO, MI, MO>(
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/comb/compose/mod.rs:44]: What is the ideal name here? `make_basic_composition`?

**Review by andrewvyrros — APPROVED** [review]

**Review by andrewvyrros — APPROVED** [review]

**Pull request merged**